File: contrib/extensions/git-tagger/hook.py
#!/usr/bin/env python3
"""
Git Tagger Extension for sley
Creates annotated git tags after version bumps
"""

import json
import subprocess
import sys


def run_git_command(args, cwd):
    """Execute a git command and return the result."""
    try:
        result = subprocess.run(
            ["git"] + args, cwd=cwd, capture_output=True, text=True, check=True
        )
        return True, result.stdout.strip()
    except subprocess.CalledProcessError as e:
        return False, e.stderr.strip()


def create_tag(version, config, project_root):
    """Create a git tag based on configuration."""
    # Extract config options with defaults
    prefix = config.get("prefix", "v")
    annotated = config.get("annotated", True)
    sign = config.get("sign", False)
    message_template = config.get("message", "Release {version}")
    push = config.get("push", False)

    # Build tag name
    tag_name = f"{prefix}{version}"

    # Build tag message
    tag_message = message_template.replace("{version}", version)

    # Check if tag already exists
    success, _ = run_git_command(["tag", "-l", tag_name], project_root)
    if success:
        existing_tags_success, existing_tags = run_git_command(
            ["tag", "-l", tag_name], project_root
        )
        if existing_tags_success and existing_tags:
            return False, f"Tag {tag_name} already exists"

    # Build git tag command
    tag_args = ["tag"]

    if annotated:
        tag_args.extend(["-a", tag_name, "-m", tag_message])
        if sign:
            tag_args.append("-s")
    else:
        tag_args.append(tag_name)

    # Create the tag
    success, output = run_git_command(tag_args, project_root)
    if not success:
        return False, f"Failed to create tag: {output}"

    # Push tag if requested
    if push:
        success, output = run_git_command(["push", "origin", tag_name], project_root)
        if not success:
            return False, f"Tag created but failed to push: {output}"
        return True, f"Created and pushed tag {tag_name}"

    return True, f"Created tag {tag_name}"


def main():
    """Main entry point for the extension."""
    try:
        # Read JSON input from stdin
        input_data = json.load(sys.stdin)

        # Extract required fields
        version = input_data.get("version")
        project_root = input_data.get("project_root")
        config = input_data.get("config", {})

        # Validate required fields
        if not version:
            result = {
                "success": False,
                "message": "Missing required field: version",
                "data": {},
            }
            print(json.dumps(result))
            sys.exit(1)

        if not project_root:
            result = {
                "success": False,
                "message": "Missing required field: project_root",
                "data": {},
            }
            print(json.dumps(result))
            sys.exit(1)

        # Create the tag
        success, message = create_tag(version, config, project_root)

        # Return result
        result = {"success": success, "message": message, "data": {}}
        print(json.dumps(result))
        sys.exit(0 if success else 1)

    except json.JSONDecodeError as e:
        result = {
            "success": False,
            "message": f"Invalid JSON input: {str(e)}",
            "data": {},
        }
        print(json.dumps(result))
        sys.exit(1)
    except Exception as e:
        result = {
            "success": False,
            "message": f"Unexpected error: {str(e)}",
            "data": {},
        }
        print(json.dumps(result))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: contrib/extensions/package-sync/hook.js
#!/usr/bin/env node
/**
 * Package Sync Extension for sley
 * Synchronizes version to package.json and other JSON files
 */

const fs = require("fs");
const path = require("path");

/**
 * Get nested property value from object using dot notation
 * @param {Object} obj - The object to traverse
 * @param {string} path - Dot-separated path (e.g., "version" or "metadata.version")
 * @returns {*} The value at the path or undefined
 */
function getNestedValue(obj, path) {
  return path.split(".").reduce((current, key) => current?.[key], obj);
}

/**
 * Set nested property value in object using dot notation
 * @param {Object} obj - The object to modify
 * @param {string} path - Dot-separated path (e.g., "version" or "metadata.version")
 * @param {*} value - The value to set
 */
function setNestedValue(obj, path, value) {
  const keys = path.split(".");
  const lastKey = keys.pop();
  const target = keys.reduce((current, key) => {
    if (!(key in current)) {
      current[key] = {};
    }
    return current[key];
  }, obj);
  target[lastKey] = value;
}

/**
 * Update a JSON file with the new version
 * @param {string} filePath - Absolute path to the JSON file
 * @param {string} version - The new version
 * @param {Array<string>} paths - JSON paths to update (default: ["version"])
 * @returns {Object} Result object with success status and message
 */
function updateJsonFile(filePath, version, paths = ["version"]) {
  try {
    // Check if file exists
    if (!fs.existsSync(filePath)) {
      return {
        success: false,
        message: `File not found: ${filePath}`,
      };
    }

    // Read and parse JSON file
    const fileContent = fs.readFileSync(filePath, "utf8");
    let jsonData;

    try {
      jsonData = JSON.parse(fileContent);
    } catch (parseError) {
      return {
        success: false,
        message: `Invalid JSON in ${filePath}: ${parseError.message}`,
      };
    }

    // Update all specified paths
    const updatedPaths = [];
    for (const jsonPath of paths) {
      const oldValue = getNestedValue(jsonData, jsonPath);
      setNestedValue(jsonData, jsonPath, version);
      updatedPaths.push(`${jsonPath}: ${oldValue} -> ${version}`);
    }

    // Write back to file with formatting
    const indentSize = detectIndentation(fileContent);
    const updatedContent = JSON.stringify(jsonData, null, indentSize) + "\n";
    fs.writeFileSync(filePath, updatedContent, "utf8");

    return {
      success: true,
      message: `Updated ${path.basename(filePath)}: ${updatedPaths.join(", ")}`,
    };
  } catch (error) {
    return {
      success: false,
      message: `Error updating ${filePath}: ${error.message}`,
    };
  }
}

/**
 * Detect indentation style of JSON file
 * @param {string} content - File content
 * @returns {number|string} Indentation (number of spaces or '\t')
 */
function detectIndentation(content) {
  // Try to detect indentation from existing content
  const match = content.match(/^(\s+)"[^"]+"/m);
  if (match) {
    const indent = match[1];
    if (indent.includes("\t")) {
      return "\t";
    }
    return indent.length;
  }
  // Default to 2 spaces
  return 2;
}

/**
 * Main entry point
 */
function main() {
  try {
    // Read JSON input from stdin
    let inputData = "";

    process.stdin.on("data", (chunk) => {
      inputData += chunk;
    });

    process.stdin.on("end", () => {
      try {
        const input = JSON.parse(inputData);

        // Extract required fields
        const version = input.version;
        const projectRoot = input.project_root;
        const config = input.config || {};

        // Validate required fields
        if (!version) {
          const result = {
            success: false,
            message: "Missing required field: version",
            data: {},
          };
          console.log(JSON.stringify(result));
          process.exit(1);
        }

        if (!projectRoot) {
          const result = {
            success: false,
            message: "Missing required field: project_root",
            data: {},
          };
          console.log(JSON.stringify(result));
          process.exit(1);
        }

        // Get file configurations with defaults
        const fileConfigs = config.files || [
          { path: "package.json", json_paths: ["version"] },
        ];

        // Process each file
        const results = [];
        let hasError = false;

        for (const fileConfig of fileConfigs) {
          const filePath = path.join(
            projectRoot,
            fileConfig.path || fileConfig,
          );
          const jsonPaths = fileConfig.json_paths || ["version"];

          const result = updateJsonFile(filePath, version, jsonPaths);
          results.push(result.message);

          if (!result.success) {
            hasError = true;
          }
        }

        // Return combined result
        const output = {
          success: !hasError,
          message: results.join("; "),
          data: {
            files_processed: fileConfigs.length,
            version: version,
          },
        };

        console.log(JSON.stringify(output));
        process.exit(hasError ? 1 : 0);
      } catch (parseError) {
        const result = {
          success: false,
          message: `Invalid JSON input: ${parseError.message}`,
          data: {},
        };
        console.log(JSON.stringify(result));
        process.exit(1);
      }
    });

    process.stdin.on("error", (error) => {
      const result = {
        success: false,
        message: `Error reading input: ${error.message}`,
        data: {},
      };
      console.log(JSON.stringify(result));
      process.exit(1);
    });
  } catch (error) {
    const result = {
      success: false,
      message: `Unexpected error: ${error.message}`,
      data: {},
    };
    console.log(JSON.stringify(result));
    process.exit(1);
  }
}

// Run main function
main();
